BedrockServer.cpp

```diff
@@ -1079,20 +1079,28 @@ void BedrockServer::worker(int threadId)
                         // Nothing to do in this case, `command->complete` will be set and we'll finish as we fall out
                         // of this block.
                     } else if (result == BedrockCore::RESULT::SERVER_NOT_LEADING) {
-                        // We won't write regardless.
+                        // Server not leading, we won't write regardless.
                         core.rollback();
 
-                        // If there are no HTTPS requests, we can just re-queue this command, otherwise, we will
+                        // If there are outstanding HTTPS requests, we can't just re-queue this command, we would
                         // potentially run the same HTTPS requests twice.
                         if (command->httpsRequests.size()) {
                             SALERT("Server stopped leading while running command with HTTPS requests!");
                             command->response.methodLine = "500 Leader stopped leading";
                             _reply(command);
                             break;
                         } else {
+                            // If we're shutting down, we can't recover from the current state changes. Abandon command.
+                            if(_shutdownState.load() != RUNNING) {
+                                SINFO("State changed before 'processCommand' and we're shutting down. Abandoning " << command->request.methodLine << ".");
+                                command->response.methodLine = "500 Leader stopped leading";
+                                command->complete = true;
+                            // If we're not shutting down, we can likely recover from the current state changes.
                             // Allow for an extra retry and start from the top, like with ABANDONED_FOR_CHECKPOINT.
-                            SINFO("State changed before 'processCommand' but no HTTPS requests so retrying.");
-                            ++retry;
+                            } else {
+                                SINFO("State changed before 'processCommand' but no HTTPS requests so retrying.");
+                                ++retry;
```

Comment:
Discussed about this with @righdforsa:

Why would we keep this `++retry;` here -- in which situation would retrying infinitely (see new line 1125) would ever benefit us? This log line hasn't been logged in over 1 week. Can we not just give this very rare flow the regular amount of retries @tylerkaraszewski ? `++retry` in `ABANDONED_FOR_CHECKPOINT` conceptually makes theoretical sense, but I just can't make any sense of it here.

Reply:
We did not count any of the exception cases (like this one, the state changed while we were running the command) toward the retry count. The purpose of the retry count is to avoid conflicts, when we run out of retries, we push the command onto the blocking commit queue. Why should this state change affect the conflict-handling behavior?

Yes, we could remove it and it would likely make little difference but we could also leave it and it would likely make little difference because it's a weird exception handling case that almost never happens.

Sending the command to the blocking commit queue is actually really weird if we have just changed from LEADING to FOLLOWING (I think it will actually still work, though).

Reply:
>we could also leave it and it would likely make little difference because it's a weird exception handling case that almost never happens.

Right, but my point is that this can result, and has resulted, in infinite loops. So if we can just avoid this possibility, we're not really making anything worse, are we?

Reply:
Are we sure this in particular has resulted in an infinite loop? I'm not actually sure how that would happen.

Reply:
@righdforsa seemed sure when he explained it to me yesterday 😅

Reply:
I did not hear that explanation. :)

Reply:
> Are we sure this in particular has resulted in an infinite loop? I'm not actually sure how that would happen.

That's the whole point of the ticket that spawned this PR. When we got into shutdown, there was a command running that happened to take just long enough that it began processing before the leader node got past standingdown but when "processCommand()" hit the "is this node leading?" block, it was no longer leading, and returned a "not leading" result. Because of the not leading result, we hit this retry increment, and the command looped over and over and never recovered, which caused the sync thread to not be able to shut down, which left the node port open, which caused peers to think it was still available, and caused a 5 minute site outage.

Reply:
Now that that condition is handled (in this PR) in order to get there again, we'd have to disconnect from the cluster but not be shutting down. I don't know if we want to loop endlessly in that case.

Reply:
> Because of the not leading result, we hit this retry increment, and the command looped over and over and never recovered

Ah, looking at it now, I see how it could do that. That's not how it's intended to work. The reason it gets stuck in the loop is because it doesn't notice it's not leading at the top of the loop.

This block of code determines if this command can write:
https://github.com/Expensify/Bedrock/blob/master/BedrockServer.cpp#L924-L946

But it's only run before the retry loop.

We could move it inside the loop, or we could just set `canWriteParallel` to false when we notice we're no longer leading.